include/odp/api/spec/stash.h

```diff
@@ -0,0 +1,288 @@
+/* Copyright (c) 2020, Nokia
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+/**
+ * @file
+ *
+ * ODP stash
+ */
+
+#ifndef ODP_API_SPEC_STASH_H_
+#define ODP_API_SPEC_STASH_H_
+#include <odp/visibility_begin.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <odp/api/std_types.h>
+
+/** @defgroup odp_stash ODP STASH
+ *
+ *  @{
+ */
+
+/**
+ * @typedef odp_stash_t
+ * Stash handle
+ */
+
+/**
+ * @def ODP_STASH_INVALID
+ * Invalid stash handle
+ */
+
+/**
+ * @def ODP_STASH_NAME_LEN
+ * Maximum stash name length in chars including null char
+ */
+
+/**
+ * Stash types
+ */
+typedef enum odp_stash_type_t {
+	/** The default stash type
+	 *
+	 *  It is implementation specific in which order odp_stash_get() calls
+	 *  return object handles from the stash. The order may be FIFO, LIFO
+	 *  or something else. Use this for the best performance when any
+	 *  particular ordering is not required.
+	 */
+	ODP_STASH_TYPE_DEFAULT = 0,
+
+	/** Stash type FIFO
+	 *
+	 *  Stash is implemented as a FIFO. A stash maintains object handle
+	 *  order of consecutive odp_stash_put() calls. Object handles stored
+	 *  first in the stash are received first by following odp_stash_get()
+	 *  calls.
+	 */
+	ODP_STASH_TYPE_FIFO
+
+} odp_stash_type_t;
+
+/**
+ * Stash operation mode
+ */
+typedef enum odp_stash_op_mode_t {
+	/** Multi-thread safe operation
+	 *
+	 *  Multiple threads operate on the stash. A stash operation
+	 *  (odp_stash_put() or odp_stash_get()) may be performed concurrently
+	 *  from multiple threads.
+	 */
+	ODP_STASH_OP_MT = 0,
+
+	/** Single thread operation
+	 *
+	 *  Multiple threads operate on the stash, but application ensures that
+	 *  a stash operation (odp_stash_put() or odp_stash_get()) is not
+	 *  performed concurrently from multiple threads.
+	 */
+	ODP_STASH_OP_ST,
+
+	/** Thread local operation
+	 *
+	 *  Only a single thread operates on the stash. Both stash operations
+	 *  (odp_stash_put() and odp_stash_get()) are always performed from the
+	 *  same thread.
+	 */
+	ODP_STASH_OP_LOCAL
+
+} odp_stash_op_mode_t;
+
+/**
+ * Stash capabilities
+ */
+typedef struct odp_stash_capability_t {
+	/** Maximum number of stashes of any type */
+	uint32_t max_stashes_any_type;
+
+	/** Maximum number of stashes of this type */
+	uint32_t max_stashes;
+
+	/** Maximum number of objects per stash
+	 *
+	 *  The value of zero means that limited only by the available
+	 *  memory size. */
+	uint64_t max_num_obj;
+
+	/** Maximum supported object handle size in bytes
+	 *
+	 *  At least 4 byte object size is always supported.
```

Comment:
Yes. This tells to the application what is the maximum handle size it can store. E.g. my implementation stores one 1B or 2B handle into uint32_t.

API demands that every implementation must support at least 4B handles. I think that less than that (max 2B or 1B handles) would make the API less useful.